druid-shell/src/text_input.rs

```diff
@@ -0,0 +1,516 @@
+
+impl TextInputToken {
+    /// A token that does not correspond to any text input.
+    pub const INVALID: TextInputToken = TextInputToken(0);
+
+    /// Create a new token; this should for the most part be called only by platform code.
+    pub fn next() -> TextInputToken {
+        static TEXT_FIELD_COUNTER: Counter = Counter::new();
+        TextInputToken(TEXT_FIELD_COUNTER.next())
+    }
+
+    /// Create a new token from a raw value.
+    pub const fn from_raw(id: u64) -> TextInputToken {
+        TextInputToken(id)
+    }
+
+    /// Get the raw value for a token.
+    pub const fn into_raw(self) -> u64 {
+        self.0
+    }
+}
+
+/// An event that indicates to the platform that some aspect of a text field may have changed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum TextInputUpdate {
+    /// Indicates the value returned by `TextInputHandler::selection` may have changed.
+    SelectionChanged,
+
+    /// Indicates the values returned by one or more of these methods may have changed:
+    /// - `TextInputHandler::hit_test_point`
+    /// - `TextInputHandler::line_range`
+    /// - `TextInputHandler::bounding_box`
+    /// - `TextInputHandler::slice_bounding_box`
+    LayoutChanged,
+
+    /// Indicates any value returned from any of the TextInputHandler methods may have changed.
+    Reset,
+}
+
+/// Corresponds to a selection of some text. This also can represent a text input caret — the blinking vertical
+/// bar where text is to be inserted. A caret is a selection with zero length, where `anchor == extent`.
+/// Indices are always expressed in UTF-8 bytes, and must be between 0 and the document length, inclusive.
+///
+/// As an example, if the input caret is at the start of the document `hello world`, we would expect both `anchor`
+/// and `extent` to be `0`. If the user holds shift and presses the right arrow key five times, we would expect the
+/// word `hello` to be selected, the `anchor` to still be `0`, and the `extent` to now be `5`.
+#[derive(Clone, Default, PartialEq, Eq, Hash)]
+pub struct TextSelection {
+    /// This is the end of the selection that stays unchanged while holding shift and pressing the arrow keys.
+    pub anchor: usize,
+    /// This is the end of the selection that moves while holding shift and pressing the arrow keys.
+    pub extent: usize,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl TextSelection {
+    /// Creates a new caret (the blinking vertical bar where text is to be inserted) at UTF-8 byte index `index`.
+    /// `index` must be a grapheme cluster boundary.
+    pub fn new_caret(index: usize) -> TextSelection {
+        TextSelection {
+            anchor: index,
+            extent: index,
+        }
+    }
+
+    /// Corresponds to the end of the selection — either the `anchor` or `extent` — that has a lesser byte index.
+    /// Because of bidirectional text, this is not necessarily "left".
+    pub fn upstream_index(&self) -> usize {
```

Comment:
I actually think that `min()` and `max()` are okay for these, but curious to hear your thoughts. I do get the idea of making the names distinct in order to make people think harder about the trickiness, but I'm not sure in this case?

Reply:
I think you're right. Switched to `min` and `max`!